### PR TITLE
feat(OverflowMenu): allow breakpoint on container width

### LINKED ISSUE #7376: OverflowMenu - Breakpoint relative to parent/container width
**Is this a new component or an extension of an existing one?**
An extension

What is the existing component, if any?
OverflowMenu

**Describe the feature**
Exactly the same as it is now but relative to container width. When I resize the container, overflowing elements are sent to the overflow menu.

Quick mockup here: [https://wireframe.cc/dv1DhR](https://wireframe.cc/dv1DhR)

### PR DESCRIPTION
**Pull request by thatblindgeye**
**What**: Closes #7376

[Breakpoint on container example](https://patternfly-react-pr-7913.surge.sh/components/overflow-menu#breakpoint-on-container)

**Additional issues**:

### CODE REVIEW

**Review by mcoker — COMMENTED** [review]
Hopefully I'm reading this correctly! If not, please disregard :D It looks like  `hasBreakpointOnContainer` sets the observer on the `.pf-c-overflow-menu` parent element width? If so, it looks like in order for that to work, the `.pf-c-overflow-menu` needs some kind of (always fluid?) width that allows it to shrink. Otherwise, by default the overflow menu is only as wide as its contents since it's an `inline-flex` box, and it won't shrink below its content width since all of the component children that are flex children are using the default `min-width: min-content` size, which keeps them and the parent from shrinking in size. So basically it never changes in size.

A use case showing an issue with that might just be the component example box. As I shrink the page down, the example container (`.ws-preview`) shrinks in size, but the overflow menu doesn't (I didn't give it a fluid width) - so watching the overflow menu's width won't help here unless it's fluid - like the example added in this PR. But you may not want it to have 100% width, like if you want to place something beside the overflow menu inline with it or whatever.

Screen Shot 2022-09-01 at 7 42 44 PM

Screen Shot 2022-09-01 at 7 42 40 PM

The ability to observe the overflow menu component's box may be a valid/useful use case, but I'm wondering if it makes more sense to watch the overflow menu's parent's container size (whatever wraps the overflow menu). Or even the parent's parent (grandparent) - implying the ability to specify which element's size to watch to trigger the overflow component. If you could specify the element, you could just specify the overflow menu if you have the use case shown in this new PR's example.

A use case where you might want to watch the grandparent is if the overflow menu's parent is a basic flex child, so it is only as wide as the overflow menu. Watching the grandparent flex container size might work better. Or maybe the grandparent is also a flex child, or has `display: contents;`, so you want to watch the great-grandparent, or this is part of a sub-grid layout, in which case who knows what parent you might want to watch.

**Comment by thatblindgeye** [comment]
@mcoker Ah that's a good point about the flex containers! I can update things to allow specifying a specific element to observe rather than only the overflow menu itself

**Review by nicolethoen — CHANGES_REQUESTED** [review]
I imagine we'd want some tests to verify this behavior

**Review by mcoker — COMMENTED** [review]
Hmm... this works great, but I think I was expecting something else though I may just be a bit confused. From the original issue:

> Exactly the same as it is now but relative to container width. When I resize the container, overflowing elements are sent to the overflow menu.

I read that as relative to container width _instead of breakpoint values_ - meaning when the container width is less than the width of the overflow items, we trigger the overflow menu to show. Though the way you've done it seems perfectly valid, too.

Curious what others think. Maybe we should tag the user that opened the issue as a reviewer?

File: packages/react-core/src/components/OverflowMenu/OverflowMenu.tsx [hunk 1]
```diff
@@ -2,69 +2,91 @@ import * as React from 'react';
...
+      if ((breakpointReference as React.RefObject<any>).current) {
+        return (breakpointReference as React.RefObject<any>).current;
+      } else if (typeof breakpointReference === 'function') {
+        return breakpointReference();
+      }
+    }
+  }
+
   componentDidMount() {
+    const reference = this.getBreakpointRef();
```
Comment (tlabaj) [inline packages/react-core/src/components/OverflowMenu/OverflowMenu.tsx:49]: should this only be called if `props.breakpointReference`

File: packages/react-core/src/components/OverflowMenu/OverflowMenu.tsx [hunk 2]
```diff
@@ -2,69 +2,91 @@ import * as React from 'react';
...
 }
 
 export class OverflowMenu extends React.Component<OverflowMenuProps, OverflowMenuState> {
   static displayName = 'OverflowMenu';
   constructor(props: OverflowMenuProps) {
     super(props);
     this.state = {
-      isBelowBreakpoint: false
+      isBelowBreakpoint: false,
+      breakpointRef: null
```
Comment (tlabaj) [inline packages/react-core/src/components/OverflowMenu/OverflowMenu.tsx:31]: can it be undefined?

**Review by tlabaj — CHANGES_REQUESTED** [review]

**Comment by thatblindgeye** [comment]
@nicolethoen @tlabaj pushed updates based on the last version of the PR, adding tests and making suggested changes.

@mcoker based on the conversation in the PR and what was discussed when we sat down to go over some local code I had wanted to show, were there any updates you wanted to see in this PR? Or would a followup issue to potentially extend the component's functionality make sense?

**Review by nicolethoen — APPROVED** [review]
I like it

**Review by tlabaj — COMMENTED** [review]
LGTM

File: packages/react-core/src/components/OverflowMenu/__tests__/OverflowMenu.test.tsx [hunk 3]
```diff
@@ -42,4 +43,41 @@ describe('OverflowMenu', () => {
 
     expect(myMock).toHaveBeenCalled();
   });
+
+  test('should call resizeObserver on undefined containerRefElement', () => {
+    const resizeObserver = jest.spyOn(getResizeObserver, 'getResizeObserver');
+
+    render(<OverflowMenu breakpoint={'md'} />);
+
+    expect(resizeObserver.mock.calls[1]).toEqual(undefined);
```
Comment (wise-king-sullyman) [inline packages/react-core/src/components/OverflowMenu/__tests__/OverflowMenu.test.tsx:52]: Nit: per our discussion on Slack I think the usage of `.mock.calls` could be made a bit more readable using the `toHaveBeenNthCalledWith` matcher. This also applies to lines 68 and 81.

**Review by wise-king-sullyman — COMMENTED** [review]
This looks solid and that demo seems really slick! I have one nit on the tests then I'd be ready to approve.

**Review by wise-king-sullyman — APPROVED** [review]
🔥

**Review by tlabaj — APPROVED** [review]

**Review by mcoker — APPROVED** [review]
This is awesome!

**Pull request merged**